compiler/rustc_hir_typeck/src/closure.rs

```diff
@@ -500,6 +523,65 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         Some(ExpectedSig { cause_span, sig })
     }
 
+    /// When an async closure is passed to a function that has a "two-part" `Fn`
+    /// and `Future` trait bound, like:
+    ///
+    /// ```rust
+    /// use std::future::Future;
+    ///
+    /// fn not_exactly_an_async_closure(_f: F)
+    /// where
+    ///     F: FnOnce(String, u32) -> Fut,
+    ///     Fut: Future<Output = i32>,
+    /// {}
+    /// ```
+    ///
+    /// The we want to be able to extract the signature to guide inference in the async
+    /// closure. We will have two projection predicates registered in this case. First,
+    /// we identify the `FnOnce<Args, Output = ?Fut>` bound, and if the output type is
+    /// an inference variable `?Fut`, we check if that is bounded by a `Future<Output = Ty>`
+    /// projection.
+    fn extract_sig_from_projection_and_future_bound(
+        &self,
+        cause_span: Option<Span>,
+        projection: ty::PolyProjectionPredicate<'tcx>,
+    ) -> Option<ExpectedSig<'tcx>> {
+        let projection = self.resolve_vars_if_possible(projection);
+
+        let arg_param_ty = projection.skip_binder().projection_term.args.type_at(1);
+        debug!(?arg_param_ty);
+
+        let ty::Tuple(input_tys) = *arg_param_ty.kind() else {
+            return None;
+        };
+
+        // If the return type is a
```

Comment:
Since this is only called for `async` closures, and not regular closures returning futures, can it ever be anything else? I guess it could already have been inferred to the anonymous future type of the async block, but not sure how that situation can occur. Kinda tempted to explicitly ICE/error on it

Reply:
This can definitely be another type, and it will (as I said in the other thread) result in a type mismatch later on. We shouldn't ICE here.

```
fn hewwo(_: impl Fn() -> BoxedFuture<'static, ()>) {}

fn test() {
  hewwo(async || {}) // type mismatch on output
}
```

Reply:
And we don't need to error here either, since we'll always have a type mismatch. We're searching through already-registered obligations in our fulfillment context, so these things are already being here, so it's not really benefitting us to handle errors during this inference.